.travis.yml

```diff
@@ -1,35 +1,111 @@
+
 
 cache:
+  apt: true
+  ccache: true
   directories:
-    - ${TRAVIS_BUILD_DIR}/cmake_3.13.3
-    - ${TRAVIS_BUILD_DIR}/build
-    - ${HOME}/.hunter
+  - "$HOME/.hunter"
+  - "${TRAVIS_BUILD_DIR}/cmake_3.13.3"
+  - "${TRAVIS_BUILD_DIR}/build"
+  - "${TRAVIS_BUILD_DIR}/deps"
+  - $HOME/.cache
+  - $HOME/.ccache
+
+matrix:
+  include:
+  - os: linux
+    dist: trusty
+    compiler: gcc
+    addons:
+      apt:
+        sources:
+        - ubuntu-toolchain-r-test
+        packages:
+        - gcc-6
+        - g++-6
+        - git
+        - python3
+        - python3-pip
+    env:
+      - _CC=gcc-6
+      - _CXX=g++-6
+
+  - os: linux
+    dist: xenial
+    compiler: clang
+    addons:
+      apt:
+        sources:
+        - llvm-toolchain-xenial-6.0
+        packages:
+        - clang-6.0
+        - git
+        - python3
+        - python3-pip
+    env:
+      - _CC=clang-6.0
+      - _CXX=clang++-6.0
+
+
+  - os: osx
+    compiler: gcc
+    env:
+      - _CC=gcc
+      - _CXX=g++
+
+  - os: osx
+    compiler: clang
+    env:
+      - _CC=clang
+      - _CXX=clang++
+
 
 install:
-  - CMAKE_SRC_BUILD_DIR="${TRAVIS_BUILD_DIR}/cmake_3.13.3"
-  - mkdir -p ${CMAKE_SRC_BUILD_DIR} && cd ${CMAKE_SRC_BUILD_DIR}
-  - if [[ "${TRAVIS_OS_NAME}" == "linux" ]]; then CMAKE_TAR_FILE='cmake-3.13.3-Linux-x86_64'; CMAKE_PATH="${CMAKE_TAR_FILE}/bin";
-    elif [[ "${TRAVIS_OS_NAME}" == "osx" ]]; then CMAKE_TAR_FILE='cmake-3.13.3-Darwin-x86_64'; CMAKE_PATH="${CMAKE_TAR_FILE}/CMake.app/Contents/bin"; fi
-  - if [[ ! -f "${CMAKE_PATH}/cmake" ]]; then curl "https://cmake.org/files/v3.13/${CMAKE_TAR_FILE}.tar.gz" | tar xz; fi
-  - export PATH="$PWD/${CMAKE_PATH}:$PATH"
-  - cd ${TRAVIS_BUILD_DIR}
-addons:
-  apt:
-    sources:
-      - ubuntu-toolchain-r-test
-    packages:
-      - gcc-7
+  # install cmake
+- |
+  CMAKE_SRC_BUILD_DIR="${TRAVIS_BUILD_DIR}/cmake_3.13.3"
+  mkdir -p ${CMAKE_SRC_BUILD_DIR} && cd ${CMAKE_SRC_BUILD_DIR}
+  if [ "${TRAVIS_OS_NAME}" = "linux" ]; then
+    CMAKE_TAR_FILE='cmake-3.13.3-Linux-x86_64'
+    CMAKE_PATH="${CMAKE_TAR_FILE}/bin"
+  elif [ "${TRAVIS_OS_NAME}" = "osx" ]; then
+    CMAKE_TAR_FILE='cmake-3.13.3-Darwin-x86_64'
+    CMAKE_PATH="${CMAKE_TAR_FILE}/CMake.app/Contents/bin"
+  fi
+  if [[ ! -f "${CMAKE_PATH}/cmake" ]]; then
+    curl "https://cmake.org/files/v3.13/${CMAKE_TAR_FILE}.tar.gz" | tar xz;
+  fi
+  export PATH="$PWD/${CMAKE_PATH}:$PATH"
+  cd ${TRAVIS_BUILD_DIR}
+
+- >
+  if [ "$TRAVIS_OS_NAME" = "linux" ]; then
+    pyenv global 3.6
+    pip3 install requests gitpython
+  elif [ "$TRAVIS_OS_NAME" = "osx" ]; then
+    HOMEBREW_NO_AUTO_UPDATE=1 brew install -q python3 git
+    pip3 install -q requests gitpython
+  fi
 
 before_script:
-  - cmake --version
-  - mkdir -p build && cd build && cmake ..
+- cmake --version
+- python -V
+- pip -V
 
-script: make
+script:
+- export CC=$_CC
+- export CXX=$_CXX
+- cmake . -Bbuild
+- cmake --build build -- -j2
+- cmake --build build --target test
```

Comment:
ctest?

Reply:
there is no difference. 
ctest must be run from `build` dir, so for this we need to cd there first. it is more clear to run `test` target